### ISSUE TITLE
Unable to build manylinux1 wheel containing capnproto

### ISSUE THREAD

**Issue by vitaly-krugl** [sandstorm-io/capnproto#350]
I am attempting to build a manylinux1 wheel from the [nupic.core](https://github.com/numenta/nupic.core) project per [PEP-513](https://www.python.org/dev/peps/pep-0513/), which defines the manylinux1 policy. I am using the docker image quay.io/pypa/manylinux1_x86_64 created by https://github.com/pypa/manylinux to build the wheel. nupic.core builds and statically links against the [capnproto library](https://github.com/sandstorm-io/capnproto). Capnproto relies on signalfd.h by default. Per manylinux1 policy, the docker image quay.io/pypa/manylinux1_x86_64 is based on CentOS-5, which does not provide  signalfd.h, so my build fails like this:

```
Linking CXX static library libkj.a
[ 27%] Built target kj
[ 29%] Building CXX object src/kj/CMakeFiles/kj-async.dir/async.c++.o
[ 30%] Building CXX object src/kj/CMakeFiles/kj-async.dir/async-unix.c++.o
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-unix.c++:36:26: fatal error: sys/signalfd.h: No such file or directory
 #include <sys/signalfd.h>
```

I even tried a capnproto-specific work-around to steer capnproto from using signalfd.h by setting `-DKJ_USE_EPOLL=0`. However, in that case, the build failed due to lack of other functions and constants in manylinux1/CentOS-5:

```
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++: In member function ‘int kj::{anonymous}::SocketAddress::socket(int) const’:
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++:348:13: error: ‘SOCK_NONBLOCK’ was not declared in this scope
     type |= SOCK_NONBLOCK | SOCK_CLOEXEC;
             ^
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++:348:29: error: ‘SOCK_CLOEXEC’ was not declared in this scope
     type |= SOCK_NONBLOCK | SOCK_CLOEXEC;
                             ^
In file included from /nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++:24:0:
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++: In lambda function:
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++:648:38: error: ‘O_CLOEXEC’ was not declared in this scope
   KJ_SYSCALL(pipe2(fds, O_NONBLOCK | O_CLOEXEC));
                                      ^
/nupic.core/build/scripts/ThirdParty/Source/CapnProto/src/kj/async-io.c++:648:47: error: ‘pipe2’ was not declared in this scope
   KJ_SYSCALL(pipe2(fds, O_NONBLOCK | O_CLOEXEC));
```

nupic.core doesn't make use of the async functionality. If capnproto provided a build flag to exclude async (async.c, async-io.c, async-unix.c, and friends) from compilation and linking, that would be one way to solve the problem (CAPNP_LITE excludes too much, so it doesn't help solve this problem)

**Comment by paulproteus** [comment]
Hi @vitaly-krugl ,

I just wanted to say that I (who works at Sandstorm) saw this bug and I find it fascinating.

FWIW, @vitaly-krugl , since you don't need the RPC mechanism, you can get away with a change that possibly breaks it. Specifically, if you globally search and replace `| SOCK_CLOEXEC` with `` (empty string), maybe capnproto will then build. You could also go sort of berserk and compile it with `-DSOCK_CLOEXEC=0` and see what happens.

I'll wait for someone more familiar with the C++ implementation to remark on how crazy this is. I'm personally very interested in seeing wheel-based users of capnproto, so I'm willing to suggest strange things to help you get there.

**Comment by paulproteus** [comment]
(I edited the previous comment a few times for clarity and accuracy. Just leaving a note to say I did so, in case you're reading this by email like I often am.)

**Comment by vitaly-krugl** [comment]
A good motivator might be to go through the effort of creating a manylinux1 wheel for the pycapnproto python package. Presently, it has to build upon installation, which is very time-consuming.

**Comment by kentonv** [comment]
I guess this must be a very old version of Linux -- O_CLOEXEC was added in 2.6.23, released in 2007. CentOS 5 appears to be from early 2007 and will reach EOL in early 2017.

While we could add a bunch of #ifdefs for this, I hesitate to clutter the code to support platforms a decade or more old.

That said, I am fine with a configure flag that skips building RPC, if that is sufficient. But I suspect O_CLOEXEC and flags like it are used in the non-RPC code as well...

**Comment by paulproteus** [comment]
For @kentonv 's context about manylinux1:

manylinux1 is a specification for how to build Linux binary artifacts that are sensible to distribute on the Python package index. So it's Python community convention that if compiled code can be built to work on CentOS 5 etc., then it can be made _vastly_ easier for users (and even developers) to install and use.

See the specification here if you like: https://www.python.org/dev/peps/pep-0513/

I would personally _love_ _love_ _love_  to see a manylinux1 wheel ("wheel" being the Python community term for "binary artifact") of capnproto.

**Comment by kentonv** [comment]
In order to support this, though, we would basically have to treat manylinux1 as a different operating system target on our supported platforms list. Linux has added a lot of really useful features in the last 10 years and I tend to make use of them in my code.

Is decade-old Linux really the only option they allow?

**Comment by vitaly-krugl** [comment]
To be kosher about it, it has to be built in their official centos-5 docker image. Otherwise, it doesn't pass their test. I am cheating for now by creating my own "manylinux" docker image that's based on centos-6.8.

By the way, it would be totally awesome to have a pycapnp manylinux wheel. The pip install/build takes too long right now and slows down continuous integration.

**Comment by vitaly-krugl** [comment]
@kentonv, I agree with your assessment about treating manylinux1 as a different operating system target for capnp.

Manylinux finally offers the ability to build a python wheel that runs on most modern x86-based linux distributions and is the only Linux binary wheel type accepted by PyPi for easy installation via pip.

However, the concept of widely-portable pre-compiled Linux binaries isn't unique; it was vetted earlier by the Anaconda project and others.

This would also enable https://github.com/jparyani/pycapnp to create a manylinux wheel (cc @jparyani), thus avoiding the lengthy install/build cycle on linux platforms. Ever since manylinux, Numpy has become a breeze to install on Linux versus the previously-lengthy compilation cycle.

**Comment by habnabit** [comment]
Nice! I found this issue trying to do this exact thing. Needless to say, :+1:.

**Comment by tjedwards** [comment]
I am able to successfully build with CXXFLAGS="-DKJ_USE_EPOLL=0 -D__BIONIC__=1" which disables epoll and builds and uses the alternate bionic code which has an alternate implementation compatible with the older rhel-5 kernel.

**Comment by kentonv** [comment]
Neat! Maybe the __BIONIC__ ifdefs should be updated to also check for the rhel-5 environment somehow?

**Comment by kentonv** [comment]
If someone wants to submit a PR which `#ifdef` checks for the manylinux environment in all the right places (e.g. the places that check for bionic currently), I'll be happy to merge it.

**Comment by timkpaine** [comment]
this issue can be closed, 2_28 is the latest many linux image and it builds fine with no modifications https://github.com/1kbgz/capnproto-pypi
